### ISSUE TITLE
xgboost after onehot : Memory error

### ISSUE THREAD

**Issue by WoNiuHu** [dmlc/xgboost#3233]
For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.

## Environment info
Operating System:
Ubuntu 14.04.5
Compiler:

Package used (python/R/jvm/C++):
python2.7.6
`xgboost` version used:
0.7
If installing from source, please provide 

1. The commit hash (`git rev-parse HEAD`)
2. Logs will be helpful (If logs are large, please upload as attachment).

If you are using jvm package, please 

1. add [jvm-packages] in the title to make it quickly be identified
2. the gcc version and distribution

If you are using python package, please provide

1. The python version and distribution
2. The command to install `xgboost` if you are not installing from source

If you are using R package, please provide

1. The R `sessionInfo()`
2. The command to install `xgboost` if you are not installing from source

## Steps to reproduce

1.df = pd.read_table('train_a.csv', engine='python', sep=",")
# shuffle data.
df = df.sample(n=train_line, replace=True, axis=0).reset_index(drop=True)
train = df
2.df_train_data = train.drop(['id', 'label'], axis=1)
df_train_target = train['label']
train_x = df_train_data
train_y = df_train_target
3.train_x = train_x.astype(str)
columns_train = []
for i in train_x.columns:
    feature = LabelEncoder().fit_transform(train_x[i]).reshape(train_x.shape[0], 1)
    onehot_encoder = OneHotEncoder(sparse=False)
    feature = onehot_encoder.fit_transform(feature)
    columns_train.append(feature)
train_x = np.column_stack(columns_train)
## What have you tried?

1. when the parameter train_lines is small, it can run normaly; but when it's a little large such as train_lines=10000, it will make error "Memory Error"
2. I also used pandas.get_dummies for onehot, it still makes the same error. 
3. How to handle the categorical features with a large data.    train_data about 70M,  the machine memory about 40G; had 413 features one sample. about 10000 samples will make error.
